sources/common/inc/SafeUInt.h

```diff
@@ -0,0 +1,221 @@
+/// @file common/inc/SafeUInt.h
+///
+/// Safe integer class template.
+///
+/// @par Full Description
+/// Safe integer class overrides arthmetic operators throwing
+/// exceptions on the overflow.
+///
+/// @ingroup Generic
+///
+/// @par Copyright (c) 2017 vcdMaker team
+///
+/// Permission is hereby granted, free of charge, to any person obtaining a
+/// copy of this software and associated documentation files (the "Software"),
+/// to deal in the Software without restriction, including without limitation
+/// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+/// and/or sell copies of the Software, and to permit persons to whom the
+/// Software is furnished to do so, subject to the following conditions:
+///
+/// The above copyright notice and this permission notice shall be included
+/// in all copies or substantial portions of the Software.
+///
+/// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+/// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+/// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+/// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+/// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+/// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+/// IN THE SOFTWARE.
+
+#pragma once
+
+/// @defgroup Generic Generic
+///
+/// @brief Generic classes.
+///
+/// @par Full Description
+/// The group contains classes which could be used anywhere else.
+
+#include <cstdint>
+#include <limits>
+#include <iostream>
+#include <stdexcept>
+
+/// A safe unsigned integer class.
+/// The overridden arthmetic operators throw exceptions on the overflow.
+template<class T> class SafeUInt
+{
+    public:
+        /// The safe integer constructor.
+        ///
+        /// @param value The value of the integer.
+        SafeUInt(T value) :
+            m_Value(value)
+        {
+        }
+
+        /// The default constructor.
+        ///
+        /// By default the integer is set to 0.
+        SafeUInt() :
+            m_Value(0)
+        {
+        }
+
+        /// Returns the integer value.
+        T GetValue() const
+        {
+            return m_Value;
+        }
+
+        /// Overrides 'equal to' operator.
+        bool operator==(const SafeUInt<T> &rhs) const
```

Comment:
Should not be class member (C.161).

Reply:
It would be best to follow:
https://stackoverflow.com/questions/4421706/what-are-the-basic-rules-and-idioms-for-operator-overloading
And/or similar guidelines.